ankaios_sdk/_components/control_interface.py

```diff
@@ -0,0 +1,365 @@
+        self._output_file = None
+        # The state of the control interface must not be changed directly.
+        # Use the change_state method instead.
+        self.state = ControlInterfaceState.DISCONNECTED
+        self._read_thread = None
+        self._disconnect_event = threading.Event()
+
+        self._add_response_callback = add_response_callback
+        self._state_changed_callback = state_changed_callback
+
+        self.logger = get_logger()
+
+    def connect(self) -> None:
+        """
+        Connect to the control interface by starting to read
+        from the input fifo and opening the output fifo.
+
+        Raises:
+            AnkaiosConnectionException: If an error occurred.
+        """
+        if self.state == ControlInterfaceState.CONNECTED:
+            raise ControlInterfaceException("Already connected.")
+
+        if not os.path.exists(
+                f"{self.ANKAIOS_CONTROL_INTERFACE_BASE_PATH}/input"):
+            raise ControlInterfaceException(
+                "Control interface input fifo does not exist."
+            )
+
+        if not os.path.exists(
+                f"{self.ANKAIOS_CONTROL_INTERFACE_BASE_PATH}/output"):
+            raise ControlInterfaceException(
+                "Control interface output fifo does not exist."
+            )
+
+        # pylint: disable=consider-using-with
+        try:
+            self._output_file = open(
+                f"{self.ANKAIOS_CONTROL_INTERFACE_BASE_PATH}/output", "ab"
+            )
+        except Exception as e:
+            self.logger.error("Error while opening output fifo: %s", e)
+            raise ControlInterfaceException(
+                "Error while opening output fifo."
+            ) from e
+
+        self._read_thread = threading.Thread(
+            target=self._read_from_control_interface,
+            daemon=True
+        )
+        self.change_state(ControlInterfaceState.CONNECTED)
+        self._read_thread.start()
+        self._send_initial_hello()
+
+    def disconnect(self) -> None:
+        """
+        Disconnect from the control interface.
+        """
+        if not self.state == ControlInterfaceState.CONNECTED:
+            self.logger.debug("Already disconnected.")
+            return
+
+        self.logger.debug("Disconnecting..")
+        self._disconnect_event.set()
+        if self._read_thread is not None:
+            self._read_thread.join(timeout=2)
+            if self._read_thread.is_alive():
+                self.logger.error("Read thread did not stop.")
+            self._read_thread = None
+        self._cleanup()
+
+    def _cleanup(self) -> None:
+        """
+        Clean up the resources.
+        """
+        self.change_state(ControlInterfaceState.DISCONNECTED)
+        if self._output_file is not None:
+            self._output_file.close()
+            self._output_file = None
+        self.logger.debug("Cleanup happened")
+
+    def change_state(self, state: ControlInterfaceState) -> None:
```

Comment:
This rule was not respected throughout the python sdk. For the Ankaios class for instance, the order is: special class methods, properties, private and then public. For the Workload class, the methods are grouped by their use case.

Reply:
I cannot find any specific rules in the pep8 guide. But different method order in every python file seems like to be non-acceptable in terms of clean code. If there is no special language rule, I suggest the vertical ordering is according to general clean code guidelines. Public, private, all in top down with caller listed above the callee.

1. Class-level attributes
2. __init__()
3. Magic methods like __str__, __len__, ...
4. Public instance methods
5. private or protected methods
6. static and class methods
7. cleanup (__del__)

This ensures consistency when reading code.

Reply:
Then I will have to modify all the sdk. Because this would be a major change, and to not overpopulate this PR, I will make a separate issue for it. Is this ok?

Reply:
Ok, then let's create a separate issue for that.

Reply:
Created a new issue for this: https://github.com/eclipse-ankaios/ank-sdk-python/issues/44
This finding can be resolved now.